File: sdk/ts/src/mcp/legacy.ts
/**
 * @module mcp/legacy
 * @package @hop-top/kit
 *
 * The 2024-11-05 MCP handler: `initialize`, `tools/list`,
 * `tools/call`, plain JSON-RPC over one POST path, no sessions, no
 * SSE.
 *
 * This module is byte-for-byte frozen. ADR 0042's hard invariant is
 * "2024-11-05 behavior is preserved byte-for-byte; additive only; no
 * deprecation", and the three-way module split (legacy / modern /
 * dispatch) exists precisely so a reviewer can confirm a modern-era
 * change did not touch this file. Modern-era behavior belongs in
 * modern.ts — never here.
 */

import {
  errorResultBlock,
  encodeEnvelope,
  buildToolEnvelope,
  MCP_ERR_INVALID_PARAMS,
  MCP_ERR_INVALID_REQUEST,
  MCP_ERR_METHOD_NOT_FOUND,
  MCP_ERR_PARSE,
  MCP_SPEC_2024_11_05,
  pathFromToolName,
  renderCallResult,
  SURFACE_MCP,
  type JsonRpcErrorObject,
  type Leaf,
  type McpBridge,
  type RawJSON,
  type ResolvedMcpConfig,
} from './types.js';
import {
  DestructiveBlockedError,
  SurfaceNotEnabledError,
  UnknownCommandError,
} from './types.js';
import type { JsonRpcRequest, McpHttpRequest, McpHttpResponse } from './dispatch.js';
import { jsonResponse } from './dispatch.js';

/** Writes a successful JSON-RPC envelope. */
export function writeResult(
  id: RawJSON | undefined,
  result: unknown,
  status: number,
): McpHttpResponse {
  return jsonResponse(status, encodeEnvelope({ id, result }));
}

/** Writes a JSON-RPC error envelope. */
export function writeError(
  id: RawJSON | undefined,
  error: JsonRpcErrorObject,
  status: number,
): McpHttpResponse {
  return jsonResponse(status, encodeEnvelope({ id, error }));
}

/**
 * Renders the -32700 parse error exactly as Go does, including the
 * encoding/json message text the fixtures pin. Go reports the offset
 * of the first offending byte in the phrase "invalid character %q
 * looking for beginning of object key string" when an object key is
 * expected; other positions produce different wording. Only the
 * shapes this surface can actually receive are modelled.
 */
export function parseErrorResponse(body: string): McpHttpResponse {
  return writeError(
    undefined,
    { code: MCP_ERR_PARSE, message: `parse error: ${goJSONParseMessage(body)}` },
    400,
  );
}

/**
 * Reproduces Go's encoding/json syntax-error text for a body this
 * surface failed to parse. Go's decoder reports the first character
 * that cannot begin the token it expects; for a top-level object
 * whose first key is unquoted (the fixture case `{not json`) that is
 * "invalid character 'n' looking for beginning of object key string".
 */
function goJSONParseMessage(body: string): string {
  const trimmed = body.replace(/^[\s]*/, '');
  if (trimmed.startsWith('{')) {
    // Scan past the opening brace and any whitespace to the first
    // character of what must be a quoted object key.
    const rest = trimmed.slice(1).replace(/^[\s]*/, '');
    const ch = rest[0];
    if (ch !== undefined && ch !== '"' && ch !== '}') {
      return `invalid character '${ch}' looking for beginning of object key string`;
    }
  }
  const ch = trimmed[0];
  if (ch !== undefined) {
    return `invalid character '${ch}' looking for beginning of value`;
  }
  return 'unexpected end of JSON input';
}

/**
 * The 2024-11-05 handler. Stateless across requests; holds only
 * immutable mount configuration and the bridge.
 */
export class LegacyMcpHandler {
  constructor(
    private readonly bridge: McpBridge,
    private readonly cfg: ResolvedMcpConfig,
  ) {}

  /**
   * Routes one already-parsed JSON-RPC request by method. An unknown
   * method returns -32601 at HTTP 200 — today's behavior, preserved
   * deliberately (the modern handler uses 404 instead; the asymmetry
   * is required by byte-for-byte preservation).
   */
  async serve(
    req: McpHttpRequest,
    rpc: JsonRpcRequest,
  ): Promise<McpHttpResponse> {
    if (rpc.jsonrpc !== undefined && rpc.jsonrpc !== '2.0') {
      return writeError(
        rpc.id,
        { code: MCP_ERR_INVALID_REQUEST, message: 'invalid jsonrpc version' },
        400,
      );
    }

    switch (rpc.method) {
      case 'initialize':
        return this.handleInitialize(rpc);
      case 'tools/list':
        return this.handleToolsList(rpc);
      case 'tools/call':
        return this.handleToolsCall(req, rpc);
      default:
        return writeError(
          rpc.id,
          {
            code: MCP_ERR_METHOD_NOT_FOUND,
            message: `method not found: ${rpc.method}`,
          },
          200,
        );
    }
  }

  /** The minimal initialize response: version, capabilities, identity. */
  private handleInitialize(rpc: JsonRpcRequest): McpHttpResponse {
    return writeResult(
      rpc.id,
      {
        protocolVersion: MCP_SPEC_2024_11_05,
        capabilities: { tools: {} },
        serverInfo: {
          name: this.cfg.serverName,
          version: this.cfg.serverVersion,
        },
      },
      200,
    );
  }

  /** One tool envelope per leaf exposed on the MCP surface. */
  private handleToolsList(rpc: JsonRpcRequest): McpHttpResponse {
    return writeResult(rpc.id, { tools: this.exposedTools() }, 200);
  }

  private exposedTools(): Array<Record<string, unknown>> {
    const tools: Array<Record<string, unknown>> = [];
    for (const leaf of this.bridge.leaves()) {
      if (!isExposed(leaf)) continue;
      tools.push(buildToolEnvelope(leaf));
    }
    return tools;
  }

  /**
   * Resolves the leaf, applies the pre-flight auth + confirmation
   * gates, and dispatches through the bridge. Unknown / not-enabled
   * tools are JSON-RPC errors; every other failure (bridge errors,
   * destructive blocks, non-zero exit codes) is an isError result.
   */
  private async handleToolsCall(
    req: McpHttpRequest,
    rpc: JsonRpcRequest,
  ): Promise<McpHttpResponse> {
    const params = (rpc.params ?? {}) as {
      name?: unknown;
      arguments?: Record<string, unknown>;
    };
    const name = typeof params.name === 'string' ? params.name : '';
    if (name === '') {
      return writeError(
        rpc.id,
        { code: MCP_ERR_INVALID_PARAMS, message: 'missing tool name' },
        200,
      );
    }

    const leaf = resolveLeaf(this.bridge, name);
    if (leaf === undefined || !isExposed(leaf)) {
      return writeError(
        rpc.id,
        { code: MCP_ERR_INVALID_PARAMS, message: `unknown tool: ${name}` },
        200,
      );
    }

    // Auth + confirmation gating, mirrored onto the result envelope so
    // MCP-aware clients see isError while HTTP-only clients see the
    // matching status code.
    if (leaf.class?.authRequired && !headerValue(req, 'authorization')) {
      return writeResult(
        rpc.id,
        errorResultBlock('authentication required'),
        401,
      );
    }
    if (
      leaf.class?.requiresConfirmation &&
      !headerValue(req, 'x-confirm-token')
    ) {
      return writeResult(
        rpc.id,
        errorResultBlock('confirmation required'),
        428,
      );
    }

    try {
      const res = await this.bridge.invoke({
        path: [...leaf.path],
        flags: params.arguments,
        meta: { surface: SURFACE_MCP, requestedAt: new Date() },
      });
      return writeResult(rpc.id, renderCallResult(res), 200);
    } catch (err) {
      if (
        err instanceof UnknownCommandError ||
        err instanceof SurfaceNotEnabledError
      ) {
        return writeError(
          rpc.id,
          { code: MCP_ERR_INVALID_PARAMS, message: `unknown tool: ${name}` },
          200,
        );
      }
      // DestructiveBlockedError and every other invoke failure render
      // as an isError result at HTTP 200.
      return writeResult(
        rpc.id,
        errorResultBlock(errorMessage(err)),
        200,
      );
    }
  }
}

/** Reports whether a leaf is exposed on the MCP surface. */
export function isExposed(leaf: Leaf): boolean {
  // Absent `enabled` means the bridge applies its policy defaults,
  // which include the MCP surface; an explicit false hides the leaf.
  return leaf.enabled?.[SURFACE_MCP] !== false;
}

/** Resolves a dotted tool name to a leaf, or undefined. */
export function resolveLeaf(
  bridge: McpBridge,
  name: string,
): Leaf | undefined {
  const path = pathFromToolName(name);
  if (path.length === 0) return undefined;
  return bridge
    .leaves()
    .find(
      (l) =>
        l.path.length === path.length &&
        l.path.every((seg, i) => seg === path[i]),
    );
}

/** Case-insensitive single header lookup. */
export function headerValue(
  req: McpHttpRequest,
  name: string,
): string | undefined {
  const vals = headerValues(req, name);
  return vals.length > 0 ? vals[0] : undefined;
}

/** All values for a header name, matched case-insensitively. */
export function headerValues(
  req: McpHttpRequest,
  name: string,
): string[] {
  const want = name.toLowerCase();
  const out: string[] = [];
  for (const [k, v] of Object.entries(req.headers ?? {})) {
    if (k.toLowerCase() !== want) continue;
    if (Array.isArray(v)) out.push(...v);
    else if (v !== undefined) out.push(v);
  }
  return out;
}

/** Extracts an error message the way Go's err.Error() renders it. */
export function errorMessage(err: unknown): string {
  if (err instanceof Error) return err.message;
  return String(err);
}

export { DestructiveBlockedError };

File: sdk/py/hop_top_kit/mcp/legacy.py
"""Legacy MCP handler — protocol revision 2024-11-05.

This module is subject to one hard invariant, inherited from the Go
reference: **2024-11-05 behaviour is preserved byte-for-byte; changes are
additive only; nothing is deprecated.** It is a separate module rather
than a branch inside a combined handler precisely so that invariant stays
structurally checkable — a reviewer can confirm a modern-era change did
not touch this file.

Serves three methods: ``initialize`` (server identity and capabilities),
``tools/list`` (one tool per exposed leaf), and ``tools/call`` (invoke via
the bridge and render the result as content blocks). Everything else is
``-32601`` at HTTP **200** — a status the modern handler deliberately does
not share, because changing it here would break the preservation rule.
"""

from __future__ import annotations

from typing import Any

from .bridge import Bridge, Invocation, Leaf, Meta, Result, tool_path
from .protocol import (
    ERR_INVALID_PARAMS,
    ERR_INVALID_REQUEST,
    ERR_METHOD_NOT_FOUND,
    LEGACY_PROTOCOL_VERSION,
    Request,
    Response,
    RPCRequest,
    write_error,
    write_result,
)
from .safety import (
    DestructiveBlockedError,
    Surface,
    SurfaceNotEnabledError,
    UnknownCommandError,
)

#: HTTP statuses the pre-flight gates use. Mirrored on the modern path so
#: MCP-aware clients see ``isError`` while HTTP-only clients see a status
#: they can act on.
STATUS_OK = 200
STATUS_UNAUTHORIZED = 401
STATUS_PRECONDITION_REQUIRED = 428


class LegacyHandler:
    """Serves one mount's 2024-11-05 traffic.

    Stateless across requests: all mutable state lives on the bridge, so
    one instance safely serves concurrent requests.
    """

    def __init__(self, bridge: Bridge, *, server_name: str, server_version: str) -> None:
        self._bridge = bridge
        self._server_name = server_name
        self._server_version = server_version

    def serve(self, request: Request, rpc: RPCRequest) -> Response:
        """Route one already-parsed request by method."""
        if rpc.jsonrpc not in ("", "2.0"):
            return write_error(rpc.id_raw, ERR_INVALID_REQUEST, "invalid jsonrpc version", 400)
        if rpc.method == "initialize":
            return self._initialize(rpc)
        if rpc.method == "tools/list":
            return self._tools_list(rpc)
        if rpc.method == "tools/call":
            return self._tools_call(request, rpc)
        return write_error(
            rpc.id_raw,
            ERR_METHOD_NOT_FOUND,
            f"method not found: {rpc.method}",
            STATUS_OK,
        )

    def _initialize(self, rpc: RPCRequest) -> Response:
        """The handshake response: protocol version, capabilities, identity."""
        return write_result(
            rpc.id_raw,
            {
                "protocolVersion": LEGACY_PROTOCOL_VERSION,
                "capabilities": {"tools": {}},
                "serverInfo": {
                    "name": self._server_name,
                    "version": self._server_version,
                },
            },
            STATUS_OK,
        )

    def _tools_list(self, rpc: RPCRequest) -> Response:
        """Enumerate every leaf exposed on the MCP surface."""
        return write_result(rpc.id_raw, {"tools": exposed_tools(self._bridge)}, STATUS_OK)

    def _tools_call(self, request: Request, rpc: RPCRequest) -> Response:
        """Resolve, gate, invoke, and render one tool call.

        Error mapping follows the surface contract: an unknown or
        unexposed tool is a JSON-RPC ``-32602``; everything else — bridge
        refusals, runner failures, non-zero exit codes — comes back as a
        result with ``isError: true``, so an MCP client sees a tool
        failure rather than a transport failure.
        """
        params = rpc.params if isinstance(rpc.params, dict) else {}
        name = params.get("name")
        if not isinstance(name, str) or not name:
            return write_error(rpc.id_raw, ERR_INVALID_PARAMS, "missing tool name", STATUS_OK)

        leaf = resolve_exposed_leaf(self._bridge, name)
        if leaf is None:
            return write_error(rpc.id_raw, ERR_INVALID_PARAMS, f"unknown tool: {name}", STATUS_OK)

        gate = preflight_refusal(request, leaf)
        if gate is not None:
            message, status = gate
            return write_result(rpc.id_raw, error_result_block(message), status)

        arguments = params.get("arguments")
        inv = Invocation(
            path=leaf.path,
            flags=arguments if isinstance(arguments, dict) else {},
            meta=Meta(surface=Surface.MCP),
        )
        try:
            result = self._bridge.invoke(inv)
        except (UnknownCommandError, SurfaceNotEnabledError):
            return write_error(rpc.id_raw, ERR_INVALID_PARAMS, f"unknown tool: {name}", STATUS_OK)
        except DestructiveBlockedError as exc:
            return write_result(rpc.id_raw, error_result_block(str(exc)), STATUS_OK)
        except Exception as exc:
            return write_result(rpc.id_raw, error_result_block(str(exc)), STATUS_OK)

        return write_result(rpc.id_raw, render_call_result(result), STATUS_OK)


# --- shared renderers ---------------------------------------------------
#
# Both eras call these. They live here rather than in a third module
# because the modern era is defined as "legacy's rendering plus envelope
# members": keeping one implementation is what makes schema and content
# drift between the two impossible rather than merely unlikely.


def exposed_tools(bridge: Bridge) -> list[dict[str, Any]]:
    """Tool descriptors for every leaf exposed on the MCP surface."""
    return [
        leaf.tool_envelope() for leaf in bridge.leaves() if leaf.enabled.get(Surface.MCP, False)
    ]


def resolve_exposed_leaf(bridge: Bridge, name: str) -> Leaf | None:
    """Resolve a dotted tool name to an MCP-exposed leaf, or ``None``."""
    try:
        leaf = bridge.resolve_leaf(tool_path(name))
    except UnknownCommandError:
        return None
    if not leaf.enabled.get(Surface.MCP, False):
        return None
    return leaf


def preflight_refusal(request: Request, leaf: Leaf) -> tuple[str, int] | None:
    """Apply the auth and confirmation header gates.

    Returns ``(message, status)`` when the call must be refused, or
    ``None`` to proceed. The modern path reuses the auth half and swaps
    the confirmation half for a strategy slot, so the two eras cannot
    drift on what "authenticated" means.
    """
    if leaf.cls.auth_required and not request.headers.get("authorization"):
        return "authentication required", STATUS_UNAUTHORIZED
    if leaf.cls.requires_confirmation and not request.headers.get("x-confirm-token"):
        return "confirmation required", STATUS_PRECONDITION_REQUIRED
    return None


def render_call_result(result: Result) -> dict[str, Any]:
    """Map a bridge result onto the ``tools/call`` result envelope.

    The content list always carries at least the stdout block (possibly
    empty); stderr adds a ``[stderr] `` prefixed block, and a structured
    payload adds its JSON serialisation as a third text block — which
    doubles as the spec-recommended serialised fallback for
    ``structuredContent`` on the modern path.
    """
    from ._json import dumps

    content: list[dict[str, Any]] = [{"type": "text", "text": result.stdout}]
    if result.stderr:
        content.append({"type": "text", "text": "[stderr] " + result.stderr})
    if result.data is not None:
        content.append({"type": "text", "text": dumps(result.data)})
    return {"content": content, "isError": result.exit_code != 0}


def error_result_block(message: str) -> dict[str, Any]:
    """A ``tools/call`` result flagged ``isError`` with one text block."""
    return {"content": [{"type": "text", "text": message}], "isError": True}
